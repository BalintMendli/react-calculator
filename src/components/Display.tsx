import React from 'react';
import styled from 'styled-components/macro';
import { transformForDisplay, transformError } from 'logic/utils';

interface DisplayProps {
  result: string;
  expression: string;
  error: string;
}

const StyledDisplay = styled.div`
  height: 80px;
  color: #fff;
  display: flex;
  flex-direction: column;
  align-items: flex-end;
  justify-content: flex-start;
  padding-bottom: 3px;
  font-size: 28px;
  line-height: 1.05;
`;

const StyledError = styled.div`
  height: 18px;
  max-width: 420px;
  padding: 2px 5px;
  margin-top: 5px;
  background-color: #a5243d;
  font-size: 12px;
  border-radius: 2px;
  overflow: hidden;
  white-space: nowrap;
  text-overflow: ellipsis;

  :empty {
    visibility: hidden;
  }
`;

const StyledExp = styled.div`
  height: 22px;
  font-size: 16px;
  display: flex;
  align-items: center;
  transition: all 150ms ease-in;
  transform-origin: top right;

  :empty {
    transform: translate(-15px, 25px) scale(1.5);
    opacity: 0;
  }
`;

const Display = ({ result, expression, error }: DisplayProps): JSX.Element => (
  <StyledDisplay id="display">
    <StyledError id="error">
      {error ? transformError(error, result) : ''}
    </StyledError>
    <StyledExp id="expression">
      {expression ? `${transformForDisplay(expression)} =` : ''}
    </StyledExp>
    <div id="result">{transformForDisplay(result)}</div>
  </StyledDisplay>
);

export default Display;
