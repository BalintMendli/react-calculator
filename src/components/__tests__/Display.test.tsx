import React from 'react';
import Display from 'components/Display';
import { shallow, ShallowWrapper } from 'enzyme';

describe('Display component', () => {
  let wrapper: ShallowWrapper;

  beforeEach(() => {
    wrapper = shallow(
      <Display
        error="test error"
        expression="test expression"
        result="test result"
      />,
    );
  });

  it('renders correctly', () => {
    expect(wrapper.debug()).toMatchSnapshot();
  });

  it('renders display div', () => {
    expect(wrapper.find('#display').length).toEqual(1);
  });

  it('renders error div', () => {
    expect(wrapper.find('#error').length).toEqual(1);
    expect(wrapper.find('#error').text()).toEqual('test error');
  });

  it('renders expression div', () => {
    expect(wrapper.find('#expression').length).toEqual(1);
    expect(wrapper.find('#expression').text()).toEqual('test expression =');
  });

  it('renders result div', () => {
    expect(wrapper.find('#result').length).toEqual(1);
    expect(wrapper.find('#result').text()).toEqual('test result');
  });
});
