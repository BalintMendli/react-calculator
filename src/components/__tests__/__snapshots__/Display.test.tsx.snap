// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Display component renders correctly 1`] = `
"<Display__StyledDisplay id=\\"display\\">
  <Display__StyledError id=\\"error\\">
    test error
  </Display__StyledError>
  <Display__StyledExp id=\\"expression\\">
    test expression =
  </Display__StyledExp>
  <div id=\\"result\\">
    test result
  </div>
</Display__StyledDisplay>"
`;
